src/TinteroLoan.sol

```diff
@@ -0,0 +1,702 @@
+    ///         ^---- ONGOING
+    ///          ^--- DEFAULTED
+    ///           ^-- REPOSSESSED
+    ///            ^- PAID
+    function _encodeStateBitmap(
+        LoanState loanState
+    ) private pure returns (bytes32) {
+        return bytes32(1 << uint8(loanState));
+    }
+
+    /// @dev Repays the payments from `start` to `end` by doing a single transfer per tranche.
+    /// Assumes end is not greater than the total number of payments.
+    ///
+    /// Requirements:
+    ///
+    /// - The sender MUST have enough funds to repay the principal of the specified payments
+    /// - The sender MUST have approved this contract to transfer the principal amount
+    ///
+    /// Effects:
+    ///
+    /// - The principal of the repaid payments is transferred from the sender to the receiver of each payment tranche
+    /// - Moves to ONGOING if paid until below the default threshold.
+    /// - Moves to PAID state if all payments are repaid.
+    /// - Emits a `RepaidPayment` event for each payment repaid.
+    function _repayByTranches(
+        uint256 start,
+        uint256 end
+    ) private returns (uint256 principalPaid) {
+        uint256 trancheIndex_ = currentTrancheIndex();
+
+        uint96 _start;
+        uint96 _end;
+        address _receiver;
+
+        for (
+            (_start, (_end, _receiver)) = (
+                start.toUint96(),
+                tranche(trancheIndex_)
+            );
+            _start < end;
+            (_start, (_end, _receiver)) = (_end, tranche(trancheIndex_++))
+        ) {
+            (uint256 toPay, uint256 principalPaidInPayment) = _prepareToPay(
+                _start,
+                Math.min(_end, end)
+            );
+            principalPaid += principalPaidInPayment;
+            lendingAsset().safeTransferFrom(msg.sender, _receiver, toPay);
+            if (_end >= end) break;
+        }
+    }
+
+    /// @dev Disposes the collateral from payments. Burns the collateral if `collateralReceiver` is the zero address.
+    ///
+    /// Requirements:
+    ///
+    /// - Each payment collateral MUST be owned by this contract.
+    ///
+    /// Effects:
+    ///
+    /// - The collateral is transferred to the collateralReceiver if provided, otherwise it is burned.
+    function _debitCollateral(
+        uint256 start,
+        uint256 end,
+        address collateralReceiver,
+        uint256 discountedPrincipal
+    ) private {
+        liquidityProvider().onDebit(discountedPrincipal);
+        for (uint256 i = start; i < end; i++) {
+            (uint256 tokenId, ) = payment(i);
+            if (collateralReceiver == address(0))
+                collateralAsset().burn(tokenId);
+            else
+                collateralAsset().safeTransferFrom(
+                    address(this),
+                    collateralReceiver,
+                    tokenId
+                );
+
+            // No need to update heldTokenIds since they can't be transferred back anymore
```

Comment:
Aren't they? I'm thinking if the receiver is the beneficiary, he might've already approved the Loan request.

It's not possible since you can't push payments anymore. We should note this to have it in mind in case we decide to create a way to extend the loan.

It may be fine with just restricting it to the manager.